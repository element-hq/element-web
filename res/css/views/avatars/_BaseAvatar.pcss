/*
Copyright 2024 New Vector Ltd.
Copyright 2015, 2016 OpenMarket Ltd

SPDX-License-Identifier: AGPL-3.0-only OR GPL-3.0-only
Please see LICENSE files in the repository root for full details.
*/

button.mx_BaseAvatar {
    /* <button> is a form element and by default it uses the user agent (browser) styling.
       We want it to inherit the font-family and line-height from its parent.
     */
    font-family: inherit;
    line-height: inherit;
}
