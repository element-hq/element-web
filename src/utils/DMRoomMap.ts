/*
Copyright 2024 New Vector Ltd.
Copyright 2016-2019 , 2021 The Matrix.org Foundation C.I.C.

SPDX-License-Identifier: AGPL-3.0-only OR GPL-3.0-only
Please see LICENSE files in the repository root for full details.
*/

import { uniq } from "lodash";
import { Room, MatrixEvent, EventType, ClientEvent, MatrixClient } from "matrix-js-sdk/src/matrix";
import { KnownMembership } from "matrix-js-sdk/src/types";
import { logger } from "matrix-js-sdk/src/logger";
import { Optional } from "matrix-events-sdk";

import { filterValidMDirect } from "./dm/filterValidMDirect";

/**
 * Class that takes a Matrix Client and flips the m.direct map
 * so the operation of mapping a room ID to which user it's a DM
 * with can be performed efficiently.
 *
 * With 'start', this can also keep itself up to date over time.
 */
export default class DMRoomMap {
    private static sharedInstance: DMRoomMap;

    // TODO: convert these to maps
    private roomToUser: { [key: string]: string } | null = null;
    private userToRooms: { [key: string]: string[] } | null = null;
    private hasSentOutPatchDirectAccountDataPatch: boolean;
    private mDirectEvent!: { [key: string]: string[] };

    public constructor(private readonly matrixClient: MatrixClient) {
        // see onAccountData
        this.hasSentOutPatchDirectAccountDataPatch = false;

        const mDirectRawContent = matrixClient.getAccountData(EventType.Direct)?.getContent() ?? {};
        this.setMDirectFromContent(mDirectRawContent);
    }

    /**
     * Makes and returns a new shared instance that can then be accessed
     * with shared(). This returned instance is not automatically started.
     */
    public static makeShared(matrixClient: MatrixClient): DMRoomMap {
        DMRoomMap.sharedInstance = new DMRoomMap(matrixClient);
        return DMRoomMap.sharedInstance;
    }

    /**
     * Set the shared instance to the instance supplied
     * Used by tests
     * @param inst the new shared instance
     */
    public static setShared(inst: DMRoomMap): void {
        DMRoomMap.sharedInstance = inst;
    }

    /**
     * Returns a shared instance of the class
     * that uses the singleton matrix client
     * The shared instance must be started before use.
     */
    public static shared(): DMRoomMap {
        return DMRoomMap.sharedInstance;
    }

    public start(): void {
        this.populateRoomToUser();
        this.matrixClient.on(ClientEvent.AccountData, this.onAccountData);
    }

    public stop(): void {
        this.matrixClient.removeListener(ClientEvent.AccountData, this.onAccountData);
    }

    /**
     * Filter m.direct content to contain only valid data and then sets it.
     * Logs if invalid m.direct content occurs.
     * {@link filterValidMDirect}
     *
     * @param content - Raw m.direct content
     */
    private setMDirectFromContent(content: unknown): void {
        const { valid, filteredContent } = filterValidMDirect(content);

        if (!valid) {
            logger.warn("Invalid m.direct content occurred", content);
        }

        this.mDirectEvent = filteredContent;
    }

    private onAccountData = (ev: MatrixEvent): void => {
        if (ev.getType() == EventType.Direct) {
            this.setMDirectFromContent(ev.getContent());
            this.userToRooms = null;
            this.roomToUser = null;
        }
    };

    /**
     * some client bug somewhere is causing some DMs to be marked
     * with ourself, not the other user. Fix it by guessing the other user and
     * modifying userToRooms
     */
    private patchUpSelfDMs(userToRooms: Record<string, string[]>): boolean {
        const myUserId = this.matrixClient.getUserId()!;
        const selfRoomIds = userToRooms[myUserId];
        if (selfRoomIds) {
            // any self-chats that should not be self-chats?
            const guessedUserIdsThatChanged = selfRoomIds
                .map((roomId) => {
                    const room = this.matrixClient.getRoom(roomId);
                    if (room) {
                        const userId = room.guessDMUserId();
                        if (userId && userId !== myUserId) {
                            return { userId, roomId };
                        }
                    }
                })
                .filter((ids) => !!ids) as { userId: string; roomId: string }[]; //filter out
            // these are actually all legit self-chats
            // bail out
            if (!guessedUserIdsThatChanged.length) {
                return false;
            }
            userToRooms[myUserId] = selfRoomIds.filter((roomId) => {
                return !guessedUserIdsThatChanged.some((ids) => ids.roomId === roomId);
            });
            guessedUserIdsThatChanged.forEach(({ userId, roomId }) => {
                const roomIds = userToRooms[userId];
                if (!roomIds) {
                    userToRooms[userId] = [roomId];
                } else {
                    roomIds.push(roomId);
                    userToRooms[userId] = uniq(roomIds);
                }
            });
            return true;
        }
        return false;
    }

    public getDMRoomsForUserId(userId: string): string[] {
        // Here, we return the empty list if there are no rooms,
        // since the number of conversations you have with this user is zero.
        return this.getUserToRooms()[userId] || [];
    }

    /**
     * Gets the DM room which the given IDs share, if any.
     * @param {string[]} ids The identifiers (user IDs and email addresses) to look for.
     * @returns {Room} The DM room which all IDs given share, or falsy if no common room.
     */
    public getDMRoomForIdentifiers(ids: string[]): Room | null {
        // TODO: [Canonical DMs] Handle lookups for email addresses.
        // For now we'll pretend we only get user IDs and end up returning nothing for email addresses

        let commonRooms = this.getDMRoomsForUserId(ids[0]);
        for (let i = 1; i < ids.length; i++) {
            const userRooms = this.getDMRoomsForUserId(ids[i]);
            commonRooms = commonRooms.filter((r) => userRooms.includes(r));
        }

        const joinedRooms = commonRooms
            .map((r) => this.matrixClient.getRoom(r))
            .filter((r) => r && r.getMyMembership() === KnownMembership.Join);

        return joinedRooms[0];
    }

    public getUserIdForRoomId(roomId: string): Optional<string> {
        if (this.roomToUser == null) {
            // we lazily populate roomToUser so you can use
            // this class just to call getDMRoomsForUserId
            // which doesn't do very much, but is a fairly
            // convenient wrapper and there's no point
            // iterating through the map if getUserIdForRoomId()
            // is never called.
            this.populateRoomToUser();
        }
        // Here, we return undefined if the room is not in the map:
        // the room ID you gave is not a DM room for any user.
        if (this.roomToUser![roomId] === undefined) {
            // no entry? if the room is an invite, look for the is_direct hint.
            const room = this.matrixClient.getRoom(roomId);
            if (room) {
                return room.getDMInviter();
            }
        }
        return this.roomToUser![roomId];
    }

    public getUniqueRoomsWithIndividuals(): { [userId: string]: Room } {
        if (!this.roomToUser) return {}; // No rooms means no map.
        // map roomToUser to valid rooms with two participants
        return Object.keys(this.roomToUser).reduce(
            (acc, roomId: string) => {
                const userId = this.getUserIdForRoomId(roomId);
                const room = this.matrixClient.getRoom(roomId);
                const hasTwoMembers = room?.getInvitedAndJoinedMemberCount() === 2;
                if (userId && room && hasTwoMembers) {
                    acc[userId] = room;
                }
                return acc;
            },
            {} as Record<string, Room>,
        );
    }

    /**
     * @returns all room Ids from m.direct
     */
    public getRoomIds(): Set<string> {
        return Object.values(this.mDirectEvent).reduce((prevRoomIds: Set<string>, roomIds: string[]): Set<string> => {
            roomIds.forEach((roomId) => prevRoomIds.add(roomId));
            return prevRoomIds;
        }, new Set<string>());
    }

    private getUserToRooms(): { [key: string]: string[] } {
        if (!this.userToRooms) {
            const userToRooms = this.mDirectEvent;
            const myUserId = this.matrixClient.getUserId()!;
            const selfDMs = userToRooms[myUserId];
            if (selfDMs?.length) {
                const neededPatching = this.patchUpSelfDMs(userToRooms);
                // to avoid multiple devices fighting to correct
                // the account data, only try to send the corrected
                // version once.
                logger.warn(`Invalid m.direct account data detected (self-chats that shouldn't be), patching it up.`);
                if (neededPatching && !this.hasSentOutPatchDirectAccountDataPatch) {
                    this.hasSentOutPatchDirectAccountDataPatch = true;
                    this.matrixClient.setAccountData(EventType.Direct, userToRooms);
                }
            }
            this.userToRooms = userToRooms;
        }
        return this.userToRooms;
    }

    private populateRoomToUser(): void {
        this.roomToUser = {};
        for (const user of Object.keys(this.getUserToRooms())) {
            for (const roomId of this.userToRooms![user]) {
                this.roomToUser[roomId] = user;
            }
        }
    }
}
