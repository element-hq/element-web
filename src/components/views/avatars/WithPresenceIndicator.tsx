/*
Copyright 2024 New Vector Ltd.
Copyright 2024 The Matrix.org Foundation C.I.C.

SPDX-License-Identifier: AGPL-3.0-only OR GPL-3.0-only
Please see LICENSE files in the repository root for full details.
*/

import React, { ReactNode, useEffect, useState } from "react";
import { ClientEvent, Room, RoomMember, RoomStateEvent, UserEvent } from "matrix-js-sdk/src/matrix";
import { Tooltip } from "@vector-im/compound-web";

import { isPresenceEnabled } from "../../../utils/presence";
import { _t } from "../../../languageHandler";
import DMRoomMap from "../../../utils/DMRoomMap";
import { getJoinedNonFunctionalMembers } from "../../../utils/room/getJoinedNonFunctionalMembers";
import { useEventEmitter } from "../../../hooks/useEventEmitter";
import { BUSY_PRESENCE_NAME } from "../rooms/PresenceLabel";

interface Props {
    room: Room;
    size: string; // CSS size
    tooltipProps?: {
        tabIndex?: number;
    };
    children: ReactNode;
}

enum Presence {
    // Note: the names here are used in CSS class names
    Online = "ONLINE",
    Away = "AWAY",
    Offline = "OFFLINE",
    Busy = "BUSY",
}

function tooltipText(variant: Presence): string {
    switch (variant) {
        case Presence.Online:
            return _t("presence|online");
        case Presence.Away:
            return _t("presence|away");
        case Presence.Offline:
            return _t("presence|offline");
        case Presence.Busy:
            return _t("presence|busy");
    }
}

function getDmMember(room: Room): RoomMember | null {
    const otherUserId = DMRoomMap.shared().getUserIdForRoomId(room.roomId);
    return otherUserId ? room.getMember(otherUserId) : null;
}

export const useDmMember = (room: Room): RoomMember | null => {
    const [dmMember, setDmMember] = useState<RoomMember | null>(getDmMember(room));
    const updateDmMember = (): void => {
        setDmMember(getDmMember(room));
    };

    useEventEmitter(room.currentState, RoomStateEvent.Members, updateDmMember);
    useEventEmitter(room.client, ClientEvent.AccountData, updateDmMember);
    useEffect(updateDmMember, [room]);

    return dmMember;
};

function getPresence(member: RoomMember | null): Presence | null {
    if (!member?.user) return null;

    const presence = member.user.presence;
    const isOnline = member.user.currentlyActive || presence === "online";
    if (BUSY_PRESENCE_NAME.matches(member.user.presence)) {
        return Presence.Busy;
    }
    if (isOnline) {
        return Presence.Online;
    }
    if (presence === "offline") {
        return Presence.Offline;
    }
    if (presence === "unavailable") {
        return Presence.Away;
    }

    return null;
}

const usePresence = (room: Room, member: RoomMember | null): Presence | null => {
    const [presence, setPresence] = useState<Presence | null>(getPresence(member));
    const updatePresence = (): void => {
        setPresence(getPresence(member));
    };

    useEventEmitter(member?.user, UserEvent.Presence, updatePresence);
    useEventEmitter(member?.user, UserEvent.CurrentlyActive, updatePresence);
    useEffect(updatePresence, [member]);

    if (getJoinedNonFunctionalMembers(room).length !== 2 || !isPresenceEnabled(room.client)) return null;
    return presence;
};

const WithPresenceIndicator: React.FC<Props> = ({ room, size, tooltipProps, children }) => {
    const dmMember = useDmMember(room);
    const presence = usePresence(room, dmMember);

    let icon: JSX.Element | undefined;
    if (presence) {
        icon = (
            <div
                tabIndex={tooltipProps?.tabIndex ?? 0}
                className={`mx_WithPresenceIndicator_icon mx_WithPresenceIndicator_icon_${presence.toLowerCase()}`}
                style={{
                    width: size,
                    height: size,
                }}
            />
        );
    }

    if (!presence) return <>{children}</>;

    return (
        <div className="mx_WithPresenceIndicator">
            {children}
            <Tooltip label={tooltipText(presence)} placement="bottom">
                {icon}
            </Tooltip>
        </div>
    );
};

export default WithPresenceIndicator;
