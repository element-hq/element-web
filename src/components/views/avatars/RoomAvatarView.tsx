/*
 * Copyright 2025 New Vector Ltd.
 *
 * SPDX-License-Identifier: AGPL-3.0-only OR GPL-3.0-only OR LicenseRef-Element-Commercial
 * Please see LICENSE files in the repository root for full details.
 */

import React, { type JSX } from "react";
import { type Room } from "matrix-js-sdk/src/matrix";
import PublicIcon from "@vector-im/compound-design-tokens/assets/web/icons/public";
import VideoIcon from "@vector-im/compound-design-tokens/assets/web/icons/video-call-solid";
import OnlineOrUnavailableIcon from "@vector-im/compound-design-tokens/assets/web/icons/presence-solid-8x8";
import OfflineIcon from "@vector-im/compound-design-tokens/assets/web/icons/presence-outline-8x8";
import BusyIcon from "@vector-im/compound-design-tokens/assets/web/icons/presence-strikethrough-8x8";
import classNames from "classnames";

import RoomAvatar from "./RoomAvatar";
import { useRoomAvatarViewModel } from "../../viewmodels/avatars/RoomAvatarViewModel";
import { _t } from "../../../languageHandler";
import { Presence } from "./WithPresenceIndicator";

interface RoomAvatarViewProps {
    /**
     * The room to display the avatar for.
     */
    room: Room;
}

/**
 * Component to display the avatar of a room.
 * Currently only 32px size is supported.
 */
export function RoomAvatarView({ room }: RoomAvatarViewProps): JSX.Element {
    const vm = useRoomAvatarViewModel(room);
    // No decoration, we just show the avatar
    if (!vm.hasDecoration) return <RoomAvatar size="32px" room={room} />;

    return (
        <div className="mx_RoomAvatarView">
            <RoomAvatar
                className={classNames("mx_RoomAvatarView_RoomAvatar", {
                    // Presence indicator and video/public icons don't have the same size
                    // We use different masks
                    mx_RoomAvatarView_RoomAvatar_icon: vm.isVideoRoom || vm.isPublic,
                    mx_RoomAvatarView_RoomAvatar_presence: Boolean(vm.presence),
                })}
                size="32px"
                room={room}
            />

            {/* If the room is a public video room, we prefer to display only the video icon */}
            {vm.isPublic && !vm.isVideoRoom && (
                <PublicIcon
                    width="16px"
                    height="16px"
                    className="mx_RoomAvatarView_icon"
                    color="var(--cpd-color-icon-tertiary)"
                    aria-label={_t("room|header|room_is_public")}
                />
            )}
            {vm.isVideoRoom && (
                <VideoIcon
                    width="16px"
                    height="16px"
                    className="mx_RoomAvatarView_icon"
                    color="var(--cpd-color-icon-tertiary)"
                    aria-label={_t("room|video_room")}
                />
            )}
            {vm.presence && <PresenceDecoration presence={vm.presence} />}
        </div>
    );
}

type PresenceDecorationProps = {
    /**
     * The presence of the user in the DM room.
     */
    presence: NonNullable<Presence>;
};

/**
 * Component to display the presence of a user in a DM room.
 */
function PresenceDecoration({ presence }: PresenceDecorationProps): JSX.Element {
    switch (presence) {
        case Presence.Online:
            return (
                <OnlineOrUnavailableIcon
                    width="8px"
                    height="8px"
                    className="mx_RoomAvatarView_PresenceDecoration"
                    color="var(--cpd-color-icon-accent-primary)"
                    aria-label={_t("presence|online")}
                />
            );
        case Presence.Away:
            return (
                <OnlineOrUnavailableIcon
                    width="8px"
                    height="8px"
                    className="mx_RoomAvatarView_PresenceDecoration"
                    color="var(--cpd-color-icon-quaternary)"
                    aria-label={_t("presence|away")}
                />
            );
        case Presence.Offline:
            return (
                <OfflineIcon
                    width="8px"
                    height="8px"
                    className="mx_RoomAvatarView_PresenceDecoration"
                    color="var(--cpd-color-icon-tertiary)"
                    aria-label={_t("presence|offline")}
                />
            );
        case Presence.Busy:
            return (
                <BusyIcon
                    width="8px"
                    height="8px"
                    className="mx_RoomAvatarView_PresenceDecoration"
                    color="var(--cpd-color-icon-tertiary)"
                    aria-label={_t("presence|busy")}
                />
            );
    }
}
