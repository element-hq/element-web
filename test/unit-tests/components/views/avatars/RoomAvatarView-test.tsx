/*
 * Copyright 2025 New Vector Ltd.
 *
 * SPDX-License-Identifier: AGPL-3.0-only OR GPL-3.0-only OR LicenseRef-Element-Commercial
 * Please see LICENSE files in the repository root for full details.
 */

import React from "react";
import { render, screen } from "jest-matrix-react";
import { mocked } from "jest-mock";

import { RoomAvatarView } from "../../../../../src/components/views/avatars/RoomAvatarView";
import { mkStubRoom, stubClient } from "../../../../test-utils";
import {
    type RoomAvatarViewState,
    useRoomAvatarViewModel,
} from "../../../../../src/components/viewmodels/avatars/RoomAvatarViewModel";
import DMRoomMap from "../../../../../src/utils/DMRoomMap";
import { Presence } from "../../../../../src/components/views/avatars/WithPresenceIndicator";

jest.mock("../../../../../src/components/viewmodels/avatars/RoomAvatarViewModel", () => ({
    useRoomAvatarViewModel: jest.fn(),
}));

describe("<RoomAvatarView />", () => {
    const matrixClient = stubClient();
    const room = mkStubRoom("roomId", "roomName", matrixClient);

    DMRoomMap.makeShared(matrixClient);
    jest.spyOn(DMRoomMap.shared(), "getUserIdForRoomId").mockReturnValue(null);

    let defaultValue: RoomAvatarViewState;

    beforeEach(() => {
        defaultValue = {
            hasDecoration: true,
            isPublic: true,
            isVideoRoom: true,
            presence: null,
        };

        mocked(useRoomAvatarViewModel).mockReturnValue(defaultValue);
    });

    it("should not render a decoration", () => {
        mocked(useRoomAvatarViewModel).mockReturnValue({ ...defaultValue, hasDecoration: false });
        const { asFragment } = render(<RoomAvatarView room={room} />);
        expect(asFragment()).toMatchSnapshot();
    });

    it("should render a video room decoration", () => {
        mocked(useRoomAvatarViewModel).mockReturnValue({ ...defaultValue, hasDecoration: true, isVideoRoom: true });
        const { asFragment } = render(<RoomAvatarView room={room} />);

        expect(screen.getByLabelText("This room is a video room")).toBeInTheDocument();
        expect(asFragment()).toMatchSnapshot();
    });

    it("should render a public room decoration", () => {
        mocked(useRoomAvatarViewModel).mockReturnValue({
            ...defaultValue,
            hasDecoration: true,
            isPublic: true,
            isVideoRoom: false,
        });
        const { asFragment } = render(<RoomAvatarView room={room} />);

        expect(screen.getByLabelText("This room is public")).toBeInTheDocument();
        expect(asFragment()).toMatchSnapshot();
    });

    it("should not render a public room decoration if the room is a video room", () => {
        mocked(useRoomAvatarViewModel).mockReturnValue({
            ...defaultValue,
            hasDecoration: true,
            isPublic: true,
            isVideoRoom: true,
        });
        render(<RoomAvatarView room={room} />);

        expect(screen.getByLabelText("This room is a video room")).toBeInTheDocument();
        expect(screen.queryByLabelText("This room is public")).toBeNull();
    });

    it.each([
        { presence: Presence.Online, label: "Online" },
        { presence: Presence.Offline, label: "Offline" },
        { presence: Presence.Busy, label: "Busy" },
        { presence: Presence.Away, label: "Away" },
    ])("should render the $presence presence", ({ presence, label }) => {
        mocked(useRoomAvatarViewModel).mockReturnValue({
            ...defaultValue,
            hasDecoration: true,
            presence,
        });
        const { asFragment } = render(<RoomAvatarView room={room} />);

        expect(screen.getByLabelText(label)).toBeInTheDocument();
        expect(asFragment()).toMatchSnapshot();
    });
});
