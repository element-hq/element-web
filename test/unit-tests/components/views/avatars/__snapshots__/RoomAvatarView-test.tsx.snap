// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RoomAvatarView /> should not render a decoration 1`] = `
<DocumentFragment>
  <span
    aria-label="Avatar"
    class="_avatar_1qbcf_8 mx_BaseAvatar"
    data-color="1"
    data-testid="avatar-img"
    data-type="round"
    style="--cpd-avatar-size: 32px;"
  >
    <img
      alt=""
      class="_image_1qbcf_41"
      data-type="round"
      height="32px"
      loading="lazy"
      referrerpolicy="no-referrer"
      src="http://this.is.a.url/avatar.url/room.png"
      width="32px"
    />
  </span>
</DocumentFragment>
`;

exports[`<RoomAvatarView /> should render a public room decoration 1`] = `
<DocumentFragment>
  <div
    class="mx_RoomAvatarView"
  >
    <span
      aria-label="Avatar"
      class="_avatar_1qbcf_8 mx_BaseAvatar mx_RoomAvatarView_RoomAvatar mx_RoomAvatarView_RoomAvatar_icon"
      data-color="1"
      data-testid="avatar-img"
      data-type="round"
      style="--cpd-avatar-size: 32px;"
    >
      <img
        alt=""
        class="_image_1qbcf_41"
        data-type="round"
        height="32px"
        loading="lazy"
        referrerpolicy="no-referrer"
        src="http://this.is.a.url/avatar.url/room.png"
        width="32px"
      />
    </span>
    <svg
      aria-label="This room is public"
      class="mx_RoomAvatarView_icon"
      color="var(--cpd-color-icon-tertiary)"
      fill="currentColor"
      height="16px"
      viewBox="0 0 24 24"
      width="16px"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M12 22a9.7 9.7 0 0 1-3.9-.788 10.1 10.1 0 0 1-3.175-2.137q-1.35-1.35-2.137-3.175A9.7 9.7 0 0 1 2 12q0-2.075.788-3.9a10.1 10.1 0 0 1 2.137-3.175q1.35-1.35 3.175-2.137A9.7 9.7 0 0 1 12 2q2.075 0 3.9.788a10.1 10.1 0 0 1 3.175 2.137q1.35 1.35 2.137 3.175A9.7 9.7 0 0 1 22 12a9.7 9.7 0 0 1-.788 3.9 10.1 10.1 0 0 1-2.137 3.175q-1.35 1.35-3.175 2.137A9.7 9.7 0 0 1 12 22m-1-2.05V18q-.825 0-1.412-.587A1.93 1.93 0 0 1 9 16v-1l-4.8-4.8q-.075.45-.138.9Q4 11.55 4 12q0 3.025 1.987 5.3T11 19.95m6.9-2.55q.5-.55.9-1.187.4-.638.662-1.326.263-.687.4-1.412Q20 12.75 20 12a7.85 7.85 0 0 0-1.363-4.475A7.7 7.7 0 0 0 15 4.6V5q0 .824-.588 1.412A1.93 1.93 0 0 1 13 7h-2v2q0 .424-.287.713A.97.97 0 0 1 10 10H8v2h6q.424 0 .713.287.287.288.287.713v3h1q.65 0 1.175.387.525.388.725 1.013"
      />
    </svg>
  </div>
</DocumentFragment>
`;

exports[`<RoomAvatarView /> should render a video room decoration 1`] = `
<DocumentFragment>
  <div
    class="mx_RoomAvatarView"
  >
    <span
      aria-label="Avatar"
      class="_avatar_1qbcf_8 mx_BaseAvatar mx_RoomAvatarView_RoomAvatar mx_RoomAvatarView_RoomAvatar_icon"
      data-color="1"
      data-testid="avatar-img"
      data-type="round"
      style="--cpd-avatar-size: 32px;"
    >
      <img
        alt=""
        class="_image_1qbcf_41"
        data-type="round"
        height="32px"
        loading="lazy"
        referrerpolicy="no-referrer"
        src="http://this.is.a.url/avatar.url/room.png"
        width="32px"
      />
    </span>
    <svg
      aria-label="This room is a video room"
      class="mx_RoomAvatarView_icon"
      color="var(--cpd-color-icon-tertiary)"
      fill="currentColor"
      height="16px"
      viewBox="0 0 24 24"
      width="16px"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M6 4h10a2 2 0 0 1 2 2v4.286l3.35-2.871a1 1 0 0 1 1.65.76v7.65a1 1 0 0 1-1.65.76L18 13.715V18a2 2 0 0 1-2 2H6a4 4 0 0 1-4-4V8a4 4 0 0 1 4-4"
      />
    </svg>
  </div>
</DocumentFragment>
`;

exports[`<RoomAvatarView /> should render the AWAY presence 1`] = `
<DocumentFragment>
  <div
    class="mx_RoomAvatarView"
  >
    <span
      aria-label="Avatar"
      class="_avatar_1qbcf_8 mx_BaseAvatar mx_RoomAvatarView_RoomAvatar mx_RoomAvatarView_RoomAvatar_icon mx_RoomAvatarView_RoomAvatar_presence"
      data-color="1"
      data-testid="avatar-img"
      data-type="round"
      style="--cpd-avatar-size: 32px;"
    >
      <img
        alt=""
        class="_image_1qbcf_41"
        data-type="round"
        height="32px"
        loading="lazy"
        referrerpolicy="no-referrer"
        src="http://this.is.a.url/avatar.url/room.png"
        width="32px"
      />
    </span>
    <svg
      aria-label="This room is a video room"
      class="mx_RoomAvatarView_icon"
      color="var(--cpd-color-icon-tertiary)"
      fill="currentColor"
      height="16px"
      viewBox="0 0 24 24"
      width="16px"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M6 4h10a2 2 0 0 1 2 2v4.286l3.35-2.871a1 1 0 0 1 1.65.76v7.65a1 1 0 0 1-1.65.76L18 13.715V18a2 2 0 0 1-2 2H6a4 4 0 0 1-4-4V8a4 4 0 0 1 4-4"
      />
    </svg>
    <svg
      aria-label="Away"
      class="mx_RoomAvatarView_PresenceDecoration"
      color="var(--cpd-color-icon-quaternary)"
      fill="currentColor"
      height="8px"
      viewBox="0 0 8 8"
      width="8px"
      xmlns="http://www.w3.org/2000/svg"
    >
      <g
        clip-path="url(#a)"
      >
        <path
          d="M8 4a4 4 0 1 1-8 0 4 4 0 0 1 8 0"
        />
      </g>
      <defs>
        <clippath
          id="a"
        >
          <path
            d="M0 0h8v8H0z"
          />
        </clippath>
      </defs>
    </svg>
  </div>
</DocumentFragment>
`;

exports[`<RoomAvatarView /> should render the BUSY presence 1`] = `
<DocumentFragment>
  <div
    class="mx_RoomAvatarView"
  >
    <span
      aria-label="Avatar"
      class="_avatar_1qbcf_8 mx_BaseAvatar mx_RoomAvatarView_RoomAvatar mx_RoomAvatarView_RoomAvatar_icon mx_RoomAvatarView_RoomAvatar_presence"
      data-color="1"
      data-testid="avatar-img"
      data-type="round"
      style="--cpd-avatar-size: 32px;"
    >
      <img
        alt=""
        class="_image_1qbcf_41"
        data-type="round"
        height="32px"
        loading="lazy"
        referrerpolicy="no-referrer"
        src="http://this.is.a.url/avatar.url/room.png"
        width="32px"
      />
    </span>
    <svg
      aria-label="This room is a video room"
      class="mx_RoomAvatarView_icon"
      color="var(--cpd-color-icon-tertiary)"
      fill="currentColor"
      height="16px"
      viewBox="0 0 24 24"
      width="16px"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M6 4h10a2 2 0 0 1 2 2v4.286l3.35-2.871a1 1 0 0 1 1.65.76v7.65a1 1 0 0 1-1.65.76L18 13.715V18a2 2 0 0 1-2 2H6a4 4 0 0 1-4-4V8a4 4 0 0 1 4-4"
      />
    </svg>
    <svg
      aria-label="Busy"
      class="mx_RoomAvatarView_PresenceDecoration"
      color="var(--cpd-color-icon-tertiary)"
      fill="currentColor"
      height="8px"
      viewBox="0 0 8 8"
      width="8px"
      xmlns="http://www.w3.org/2000/svg"
    >
      <g
        clip-path="url(#a)"
      >
        <path
          clip-rule="evenodd"
          d="M8 4a4 4 0 1 1-8 0 4 4 0 0 1 8 0M5.435 6.048A2.5 2.5 0 0 1 1.687 3.05zm.914-1.19L2.648 1.897a2.5 2.5 0 0 1 3.701 2.961"
          fill-rule="evenodd"
        />
      </g>
      <defs>
        <clippath
          id="a"
        >
          <path
            d="M0 0h8v8H0z"
          />
        </clippath>
      </defs>
    </svg>
  </div>
</DocumentFragment>
`;

exports[`<RoomAvatarView /> should render the OFFLINE presence 1`] = `
<DocumentFragment>
  <div
    class="mx_RoomAvatarView"
  >
    <span
      aria-label="Avatar"
      class="_avatar_1qbcf_8 mx_BaseAvatar mx_RoomAvatarView_RoomAvatar mx_RoomAvatarView_RoomAvatar_icon mx_RoomAvatarView_RoomAvatar_presence"
      data-color="1"
      data-testid="avatar-img"
      data-type="round"
      style="--cpd-avatar-size: 32px;"
    >
      <img
        alt=""
        class="_image_1qbcf_41"
        data-type="round"
        height="32px"
        loading="lazy"
        referrerpolicy="no-referrer"
        src="http://this.is.a.url/avatar.url/room.png"
        width="32px"
      />
    </span>
    <svg
      aria-label="This room is a video room"
      class="mx_RoomAvatarView_icon"
      color="var(--cpd-color-icon-tertiary)"
      fill="currentColor"
      height="16px"
      viewBox="0 0 24 24"
      width="16px"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M6 4h10a2 2 0 0 1 2 2v4.286l3.35-2.871a1 1 0 0 1 1.65.76v7.65a1 1 0 0 1-1.65.76L18 13.715V18a2 2 0 0 1-2 2H6a4 4 0 0 1-4-4V8a4 4 0 0 1 4-4"
      />
    </svg>
    <svg
      aria-label="Offline"
      class="mx_RoomAvatarView_PresenceDecoration"
      color="var(--cpd-color-icon-tertiary)"
      fill="currentColor"
      height="8px"
      viewBox="0 0 8 8"
      width="8px"
      xmlns="http://www.w3.org/2000/svg"
    >
      <g
        clip-path="url(#a)"
      >
        <path
          clip-rule="evenodd"
          d="M4 6.5a2.5 2.5 0 1 0 0-5 2.5 2.5 0 0 0 0 5M4 8a4 4 0 1 0 0-8 4 4 0 0 0 0 8"
          fill-rule="evenodd"
        />
      </g>
      <defs>
        <clippath
          id="a"
        >
          <path
            d="M0 0h8v8H0z"
          />
        </clippath>
      </defs>
    </svg>
  </div>
</DocumentFragment>
`;

exports[`<RoomAvatarView /> should render the ONLINE presence 1`] = `
<DocumentFragment>
  <div
    class="mx_RoomAvatarView"
  >
    <span
      aria-label="Avatar"
      class="_avatar_1qbcf_8 mx_BaseAvatar mx_RoomAvatarView_RoomAvatar mx_RoomAvatarView_RoomAvatar_icon mx_RoomAvatarView_RoomAvatar_presence"
      data-color="1"
      data-testid="avatar-img"
      data-type="round"
      style="--cpd-avatar-size: 32px;"
    >
      <img
        alt=""
        class="_image_1qbcf_41"
        data-type="round"
        height="32px"
        loading="lazy"
        referrerpolicy="no-referrer"
        src="http://this.is.a.url/avatar.url/room.png"
        width="32px"
      />
    </span>
    <svg
      aria-label="This room is a video room"
      class="mx_RoomAvatarView_icon"
      color="var(--cpd-color-icon-tertiary)"
      fill="currentColor"
      height="16px"
      viewBox="0 0 24 24"
      width="16px"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M6 4h10a2 2 0 0 1 2 2v4.286l3.35-2.871a1 1 0 0 1 1.65.76v7.65a1 1 0 0 1-1.65.76L18 13.715V18a2 2 0 0 1-2 2H6a4 4 0 0 1-4-4V8a4 4 0 0 1 4-4"
      />
    </svg>
    <svg
      aria-label="Online"
      class="mx_RoomAvatarView_PresenceDecoration"
      color="var(--cpd-color-icon-accent-primary)"
      fill="currentColor"
      height="8px"
      viewBox="0 0 8 8"
      width="8px"
      xmlns="http://www.w3.org/2000/svg"
    >
      <g
        clip-path="url(#a)"
      >
        <path
          d="M8 4a4 4 0 1 1-8 0 4 4 0 0 1 8 0"
        />
      </g>
      <defs>
        <clippath
          id="a"
        >
          <path
            d="M0 0h8v8H0z"
          />
        </clippath>
      </defs>
    </svg>
  </div>
</DocumentFragment>
`;
